subprojects/core/src/test/groovy/org/gradle/api/services/internal/DefaultBuildServicesRegistryTest.groovy

```diff
@@ -356,12 +360,126 @@ class DefaultBuildServicesRegistryTest extends Specification {
         bean.getHello().getOrNull() == null
 
         when:
-        registry.registerIfAbsent("helloService", NoParamsServiceImpl) {}
+        registerService("helloService", NoParamsServiceImpl)
 
         then:
         bean.getHello().getOrNull() != null
     }
 
+    def "stop action is called when the service is not created"() {
+        given:
+        Consumer<Object> stopAction = Mock()
+        def provider = registerService("service", ServiceImpl)
+        provider.beforeStopping(stopAction)
+
+        when:
+        buildFinished()
+
+        then:
+        1 * stopAction.accept(provider)
+    }
+
+    def "service may be created by the stop action"() {
+        given:
+        def provider = registerService("service", ServiceImpl)
+        provider.beforeStopping { provider.get() }
+
+        when:
+        buildFinished()
+
+        then:
+        ServiceImpl.instances.size() == 1
+    }
+
+    def "service created by the stop action is stopped"() {
+        given:
+        def provider = registerService("service", StoppableServiceImpl)
+        provider.beforeStopping { provider.get() }
+
+        when:
+        buildFinished()
+
+        then:
+        ServiceImpl.instances.isEmpty()
+    }
+
+    def "an exception from stop action is rethrown"() {
+        given:
+        def provider = registerService("service", ServiceImpl)
+        provider.beforeStopping { throw new UnsupportedOperationException("Don't stop") }
+        when:
+        buildFinished()
+
+        then:
+        thrown(Exception)
+    }
+
+    def "other stop actions run if one of them throws"() {
```

Comment:
❓ What about other service providers also registered as listeners, would they be properly stopped as well?

Reply:
>What about other service providers also registered as listeners, would they be properly stopped as well?

I've added some tests to cover this scenario too.